### ISSUE TITLE
Error destroying App Service Plan when switching Azure Functions from Basic/Standard/Premium to Dynamic

### ISSUE THREAD

**Issue by jjgriff93** [hashicorp/terraform-provider-azurerm#5990]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

Terraform version: `v0.12.18`
AzureRM provider version: `1.40.0`

### Affected Resource(s)

* `azurerm_function_app`
* `azurerm_app_service_plan`

### Terraform Configuration Files

Initial Terraform Plan
```hcl
resource "azurerm_app_service_plan" "trafficdata" {
  name                = "myTrafficFunctionPlan"
  resource_group_name =  "myTFResourceGroup"
  location            =  "westeurope"
  kind                =  "linux"
  reserved            =  true

  sku {
    tier = "Basic"
    size = "B1"
  }
}

resource "azurerm_function_app" "trafficdata" {
  name                      = "myTrafficFinction"
  resource_group_name       = "myTFResourceGroup"
  location                  = "westeurope"
  app_service_plan_id       = azurerm_app_service_plan.trafficdata.id
  version                   = "~2"
  storage_connection_string = "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX"
  app_settings              = {
      "AZURE_MAPS_ENDPOINT"             = "https://atlas.microsoft.com/"
  }
}
```

Updated to:
```hcl
resource "azurerm_app_service_plan" "trafficdata" {
  name                = "myTrafficFunctionPlan"
  resource_group_name =  "myTFResourceGroup"
  location            =  "westeurope"
  kind                =  "functionapp"
  reserved            =  true

  sku {
    tier = "Dynamic"
    size = "Y1"
  }
}

resource "azurerm_function_app" "trafficdata" {
  name                      = "myTrafficFinction"
  resource_group_name       = "myTFResourceGroup"
  location                  = "westeurope"
  app_service_plan_id       = azurerm_app_service_plan.trafficdata.id
  version                   = "~2"
  storage_connection_string = "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX"
  app_settings              = {
      "AZURE_MAPS_ENDPOINT"             = "https://atlas.microsoft.com/"
  }
}
```

### Debug Output

https://gist.github.com/jjgriff93/cc6d3f8a383396f89583c6f960de1c45

### Expected Behavior

When changing app plan type for Azure Functions from Dynamic to Basic/Standard/Premium or vice versa, the function should first be destroyed before the app service plan, so that the new plan can be created and then a new function deployed.

### Actual Behavior

An error occurs when changing the plan type between dynamic (consumption) to app service (Basic/Standard or Premium tiers). From the error ([see gist](https://gist.github.com/jjgriff93/cc6d3f8a383396f89583c6f960de1c45)), it seems it is generated because Terraform is attempting to delete the dynamic plan first before deleting the function deployed on top of it, which needs to be deleted first before the plan is able to be deleted. 

### Steps to Reproduce

1. Create a basic Terraform script to deploy an `azurerm_app_service_plan` with the kind set to `linux` and sku set to `sku {
    tier = "Basic"
    size = "B1"
  }` and a vanilla `azurerm_function_app`
2. `terraform plan`
3. `terraform apply`
4. Change the Terraform script `azurerm_app_service_plan` kind to `functionapp` and sku to `sku {
    tier = "Basic"
    size = "B1"
  }`
5. `terraform plan`
6. `terraform apply`

* #0000

**Comment by brandonh-msft** [comment]
yes. tf isn't recognizing that in order to destroy an app service plan, all apps associated w/ it must be destroyed _first_

**Comment by ricohomewood** [comment]
Also came across this frustrating issue. Is there any ETA on a fix?

**Comment by cedricfortin** [comment]
Also afected by the issue.

I had to mark the function app as taint to force terraform to destroy my function app to be able to apply the changes:
`terraform taint azurerm_function_app.my_function_app`

**Comment by GarethOates** [comment]
Yeah this happens for any change to an app service plan that causes it to be replaced.  It doesn't recognise that the app must be replaced and instead tries to update it in place, which fails subsequently because the app_service_plan resource doesn't get recreated.

**Comment by ritugoyal12** [comment]
Getting the same error when trying to update from consumption  app service plan to premium plan. Any updates on this request ?

**Comment by lawrencegripper** [comment]
I found a workaround for this issue. By adding the "tier" to a "forceNew" property on the function and plan resources you can ensure terraform does stuff in the right order. 

```

locals {
  plan_tier    = "Premium"
  plan_size    = "EP1"
  plan_kind    = "elastic"

  // Plan and function name contains the tier and size to correctly trigger destroying of it's functions
  // when these property changes. This works as the name fields are "ForceNew" this ensures the destory and recreation of funcs/plans
  // when the kind or tier is changed. 
  // This works around the bug here: https://github.com/terraform-providers/terraform-provider-azurerm/issues/5990
  plan_name     = "${var.function_name}-${local.plan_tier}${local.plan_kind}${random_string.random.result}"
  function_name = "${var.function_name}-${local.plan_tier}${local.plan_kind}${random_string.random.result}"
}


resource "azurerm_app_service_plan" "asp" {
  name                = local.plan_name
  resource_group_name = var.shared_env.rg.name
  location            = var.shared_env.rg.location
  kind                = local.plan_kind
  sku {
    tier = local.plan_tier
    size = local.plan_size
  }
}

resource "azurerm_function_app" "functions" {
  name                       = local.function_name
  location                   = var.shared_env.rg.location
  resource_group_name        = var.shared_env.rg.name
  app_service_plan_id        = azurerm_app_service_plan.asp.id
  storage_account_name       = var.storage_account_name
  storage_account_access_key = var.storage_account_key
  version                    = "~3"
 ... more stuff here
```

**Comment by brandonh-msft** [comment]
But this changes the resource identifier of the function app which cascades to upstream (and external) dependencies, right?

**Comment by lawrencegripper** [comment]
Do you mean as it changes the name of the function end users calling an HTTP trigger would have to pickup the new name?

In our case we're ok as we output the function's URL from the TF script and all external dependencies use this to pickup the endpoint they should send requests to: 

```tf
output "function_details" {
  sensitive = false
  value = {
    defaulthostname = azurerm_function_app.functions.default_hostname
    name            = azurerm_function_app.functions.name
    resourcegroup   = azurerm_function_app.functions.resource_group_name
    baseuri         = "https://${azurerm_function_app.functions.default_hostname}"
  }
}
```
I think it would be doable without the name change to the function if `app_service_plan_id` was marked as `ForceNew` (which it feels like it should be as I'm not sure you can move a function between app service plans once it's created)

**Comment by Vasi-Shche** [comment]
I added this two things and solved this problem. Moreover, I found it very convenient that you can change the plan without deleting the function.
```hcl
resource "random_string" "this" {
  ...
  keepers = {
    kind = func_app.kind
    tier = func_app.sku_tier
    size = func_app.sku_size
  }
  ...
}

resource "azurerm_app_service_plan" "this" {
  ...
  lifecycle {
    create_before_destroy = true
  }
  depends_on = [
    random_string.this,
  ]
}
```

**Comment by drcircuit** [comment]
Any status on this? Should be a relatively easy fix? This really makes it hard to manage function apps through automation. The hack to change the name of the function app causes many issues. The most alarming one is it breaks deployment of the code through release pipelines which are tied to the id of the function app or the function app name. This bug causes manual work to have to be done in order to change something basic as a an app service plan.

**Comment by Elip-tdlr** [comment]
I too am in need of a resolution for this issue. When the service plan is deleted or replaced function apps should be deleted first. It's impossible to manage functions apps with CI/CD the way it is now.

**Comment by swetha-malyavantham-mcg** [comment]
The work around is not very reliable.. We would appreciate a fix. Is there any update on this?

**Comment by bigwheels490** [comment]
I had a similar issue where I needed to recreate an app service plan for an app service.  I solved it by applying the `create_before_destroy` lifecycle argument to the app service plan.  This forced the new app service plan to be created and the app service to be updated before destroying the old app service plan.

**Comment by chtourou-youssef** [comment]
any solution to that please ?

**Comment by rcskosir** [comment]
Thanks for opening this issue. This was a problem in the 1.x version of the provider which is no longer actively maintained.
If this is still an issue with the 4.x version of the provider please do let us know by opening a new issue, thanks!
